jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java

```diff
@@ -65,47 +67,61 @@ public class Jcmd {
+			System.out.printf(HELPTEXT);
+			return;
+		}
+
+		List<AttachProvider> providers = AttachProvider.providers();
+		if ((providers == null) || providers.isEmpty()) {
+			System.err.println("no attach providers available"); //$NON-NLS-1$
+			return;
+		}
+
+		AttachProvider openj9Provider = providers.get(0);
+		List<VirtualMachineDescriptor> vmds = openj9Provider.listVirtualMachines();
+		if ((vmds == null) || vmds.isEmpty()) {
+			System.err.println("no vm found"); //$NON-NLS-1$
+			return;
+		}
+
+		/* An empty argument list is a request for a list of VMs. */
 		final String firstArgument = (0 == args.length) ? "-l" : args[0]; //$NON-NLS-1$
 		if ("-l".equals(firstArgument)) { //$NON-NLS-1$
-			List<AttachProvider> providers = AttachProvider.providers();
-			AttachProvider myProvider = null;
-			if (!providers.isEmpty()) {
-				myProvider = providers.get(0);
-			}
-			if (null == myProvider) {
-				System.err.println("no attach providers available"); //$NON-NLS-1$
-			} else {
-				for (VirtualMachineDescriptor vmd : myProvider.listVirtualMachines()) {
-					StringBuilder outputBuffer = new StringBuilder(vmd.id());
-					Util.getTargetInformation(myProvider, vmd, false, false, true, outputBuffer);
-					System.out.println(outputBuffer.toString());
-				}
+			for (VirtualMachineDescriptor vmd : vmds) {
+				StringBuilder outputBuffer = new StringBuilder(vmd.id());
+				Util.getTargetInformation(openj9Provider, vmd, false, false, true, outputBuffer);
+				System.out.println(outputBuffer.toString());
 			}
-		} else if ((args.length == 1) && (null != firstArgument) && Arrays.stream(HELP_OPTIONS).anyMatch(firstArgument::equals)) {	
-			System.out.printf(HELPTEXT);
 		} else {
-			command = DiagnosticUtils.makeJcmdCommand(args, 1);
+			String command = DiagnosticUtils.makeJcmdCommand(args, 1);
 			if (command.isEmpty()) {
 				System.err.printf("There is no jcmd command.%n"); //$NON-NLS-1$
 				System.out.printf(HELPTEXT);
 			} else {
-				String vmid = firstArgument;
 				AttacherDiagnosticsProvider diagProvider = new AttacherDiagnosticsProvider();
-				try {
-					diagProvider.attach(vmid);
+				List<String> vmids = Util.findMatchVMIDs(vmds, firstArgument);
+				for (String vmid : vmids) {
+					if (!vmid.equals(firstArgument)) {
+						// skip following if firstArgument is a VMID
+						// keep the output compatible with the old behavior (only one VMID accepted)
+						System.out.println(vmid + ":"); //$NON-NLS-1$
+					}
 					try {
-						Util.runCommandAndPrintResult(diagProvider, command, command); // $NON-NLS-1$
-					} finally {
-						diagProvider.detach();
+						IPC.logMessage("attaching vmid = " + vmid); //$NON-NLS-1$
+						diagProvider.attach(vmid);
+						try {
+							Util.runCommandAndPrintResult(diagProvider, command, command);
+						} finally {
+							diagProvider.detach();
+						}
+					} catch (IOException e) {
+						Util.handleCommandException(vmid, e);
+						System.out.printf(HELPTEXT);
+						break;
```

Comment:
This doesn't seem like the desired behavior. If we're running the command for multiple VMs, a failure of one (perhaps because it has exited) does not mean we should skip the remaining VMs.

Reply:
Removed `break` to keep iterating the rest of VMIDs.

Reply:
I think we only want to print the help once, perhaps after the loop.

Reply:
Added a boolean `exceptionThrown` to just print the help once.
Missed this comment earlier.